[Response Ops][Task Manager ] Adding ability for ad-hoc task instance to specify timeout override

ymao1 opened the pull request:
Resolves https://github.com/elastic/kibana/issues/174353

## Summary

Adds ability for task instance to specify a timeout override that will be used in place of the task type timeout when running an ad-hoc task. In the future we may consider allowing timeout overrides for recurring tasks but this PR limits usage to only ad-hoc task runs.

This timeout override is planned for use by backfill rule execution tasks so the only usages in this PR are in the functional tests.

ymao1 commented:
/ci

ymao1 commented:
/ci

ymao1 commented:
/ci

ymao1 commented:
/ci

ymao1 commented:
/ci

elasticmachine commented:
Pinging @elastic/response-ops (Team:ResponseOps)

doakalexi reviewed (approved):
LGTM!

pmuellr reviewed (approved):
LGTM, but am wondering about the validations throwing an error.  I think it's probably fine - presumably these would get caught in development.  Given the validations - that it's a valid interval format and that it's not a recurring task - hard to imagine a "surprise" slipping through.  But figured I'd mention it since we are validating right before running - presumably such invalidated tasks just won't be run at all, vs perhaps running with the existing timeout.

Also wondering if the validation error is a ZDT thing - like would an old Kibana somehow "skip" one of these - but I don't think it will.  Older Kibanas just don't know about this at all, and it's not saved in the SO in any case.

It appears we will start persisting this field to the SO, given the code below.  Intentional?  Feels like it's too easy for us to slip fields in here! https://github.com/elastic/kibana/blob/d676f169f0486f70bdaf4d7e7a0916046e8aa83d/x-pack/plugins/task_manager/server/task_store.ts#L551-L563

ymao1 commented:
> It appears we will start persisting this field to the SO, given the code below.

I think we have to persist the timeout override since it is specified per task instance, so it needs to be in the task instance that task manager claims and reads in before running. 

That does bring up a good point about ZDT though. It looks like for rule saved objects, we're specifying a `modelVersion` schema, but we're not doing it for task manager or even connector SOs right now. Do we need to add the model versions before merging this PR?

cc @mikecote

ymao1 commented:
> I'd mention it since we are validating right before running - presumably such invalidated tasks just won't be run at all, vs perhaps running with the existing timeout.

That's a good idea, to just run with the existing timeout if the specific timeout override doesn't validate!

mikecote commented:
> That does bring up a good point about ZDT though. It looks like for rule saved objects, we're specifying a modelVersion schema, but we're not doing it for task manager or even connector SOs right now. Do we need to add the model versions before merging this PR?

I think since we switched to dynamic SO mapping and removed some fields from the mapping we're in a state where we no longer validate what goes into the SO. It would be good to add something back to ensure we have a strict schema again. Not sure if I'd see this as a blocker for this PR to merge but it would be good to get those validations in soon. Maybe we create issues and get them prioritized for 8.14?

ymao1 commented:
I guess since this PR is adding the field and the knowledge of how to use it without actually scheduling any tasks with the new field, this would fall into the bucket of "intermediate release" and by the time we would release anything that used the field any rollback would rollback to a Kibana instance that already knows about the field. Am I thinking about that right?

ymao1 commented:
@pmuellr Updated to remove `timeoutOverride` in the case of validation error so that the task uses the default task type timeout but still runs. Will log a warning. https://github.com/elastic/kibana/pull/175731/commits/30292844479994802686bd02f0a81eb3ed39e274

ymao1 commented:
> I think since we switched to dynamic SO mapping and removed some fields from the mapping we're in a state where we no longer validate what goes into the SO. It would be good to add something back to ensure we have a strict schema again. Not sure if I'd see this as a blocker for this PR to merge but it would be good to get those validations in soon. Maybe we create issues and get them prioritized for 8.14?

Created https://github.com/elastic/kibana/issues/177059
